Midichlorians/Server/server.h

```diff
@@ -0,0 +1,229 @@
+      q = &post_q;
+    } else {
+      return false;
+    }
+
+    Variant<T_WEB_EVENTS> web_event;
+    try {
+      if (q->at("ea") == "En") {
+        web_event = WebEnterEvent();
+      } else if (q->at("ea") == "Ex") {
+        web_event = WebExitEvent();
+      } else if (q->at("ea") == "Fg") {
+        web_event = WebForegroundEvent();
+      } else if (q->at("ea") == "Bg") {
+        web_event = WebBackgroundEvent();
+      } else {
+        web_event = WebGenericEvent(q->at("ec"), q->at("ea"));
+      }
+    } catch (const std::out_of_range&) {
+      return false;
+    }
+
+    if (!ExtractWebBaseEventFields(q, h, web_event)) {
+      return false;
+    }
+
+    PassLogEntryToConsumer(EventLogEntry(current::time::Now(), std::move(web_event)));
+    return true;
+  }
+
+  template <typename T_LOG_ENTRY>
+  void PassLogEntryToConsumer(T_LOG_ENTRY&& entry, bool is_valid_entry = true) {
+    T_LOG_ENTRY_VARIANT entry_variant(std::move(entry));
+    entry_variant.Call(log_entry_consumer_);
+    if (is_valid_entry) {
+      ++events_pushed_;
+      last_event_t_ = Value<LogEntryBase>(entry_variant).server_us;
+    }
+  }
+
+  bool ExtractWebBaseEventFields(const std::map<std::string, std::string>* q,
+                                 const std::map<std::string, std::string>* h,
+                                 Variant<T_WEB_EVENTS>& web_event) {
+    using namespace current::midichlorians::web;
+
+    assert(q);
+    assert(h);
+
+    WebBaseEvent& dest_event = Value<WebBaseEvent>(web_event);
+    try {
+      dest_event.user_ms = std::chrono::milliseconds(strings::FromString<uint64_t>(q->at("_t")));
+      dest_event.customer_id = q->at("CUSTOMER_ACCOUNT");
+      dest_event.client_id = q->at("cid");
+      dest_event.ip = h->at("X-Forwarded-For");
+      dest_event.user_agent = h->at("User-Agent");
+      blocks::URL url = blocks::URL(h->at("Referer"));
+      dest_event.referer_host = url.host;
+      dest_event.referer_path = url.path;
+      dest_event.referer_querystring = url.query.AsImmutableMap();
+    } catch (const std::out_of_range&) {
+      return false;
+    } catch (const blocks::EmptyURLException&) {
+      return false;
+    }
+    return true;
+  }
+
+  void LogInvalidRequest(const Request& r) {
+    PassLogEntryToConsumer(InvalidLogEntry(current::time::Now(), r), false);
+  }
+
+  void TimerThreadFunction() {
+    while (send_ticks_) {
+      std::unique_lock<std::mutex> lock(mutex_);
+      const std::chrono::microseconds now = current::time::Now();
+      const std::chrono::microseconds dt = now - last_event_t_;
+      if (dt >= tick_interval_us_) {
+        PassLogEntryToConsumer(TickLogEntry(now));
+      } else {
+        lock.unlock();
+        std::this_thread::sleep_for(tick_interval_us_ - dt + std::chrono::microseconds(1));
+      }
+    }
+  }
+
+ private:
+  std::mutex mutex_;
+  const int http_port_;
+  LOG_ENTRY_CONSUMER& log_entry_consumer_;
+  const std::string route_;
+  const std::string response_text_;
```

Comment:
This [ugly] state could go away, as well as the `ProcessEventAs...` function prefix comment above, if each processing is done within its own instance of a helper class. What do you think?